Highlight suspicious javascript functions

BugBountyzip opened the pull request:
### Bambda Contributions

* [x] Bambda has a valid [header](https://github.com/PortSwigger/bambdas/blob/73077e7ff3f6fac9db7dc95c0a00bd842b6bb64c/Proxy/HTTP/FilterOnCookieValue.bambda#L1-L5), featuring an `@author` annotation and suitable description
* [x] Bambda compiles and executes as expected
* [x] Only .bambda files have been added or modified (README.md files are automatically updated / generated after PR merge)

Hannah-PortSwigger commented:
Thanks for making this submission! Unfortunately, we've run out of time to continue this review this week, so we will provide feedback on your pull request on Monday.

ps-porpoise commented on Proxy/HTTP/DetectSuspiciousJSFunctions.bambda:
```diff
@@ -0,0 +1,72 @@
+boolean manualColorHighlightEnabled = true;
+
+// Ensure there is a response and it is not null
+if (!requestResponse.hasResponse() || requestResponse.response() == null) {
```
You can remove the `|| requestResponse.response() == null` from here - the `hasResponse()` already checks this.

BugBountyzip replied:
done

BugBountyzip replied:
done

ps-porpoise commented on Proxy/HTTP/DetectSuspiciousJSFunctions.bambda:
```diff
@@ -0,0 +1,72 @@
+            if (notesBuilder.length() > 0) {
+                notesBuilder.append(", ");
+            }
+            notesBuilder.append(function.split("\\\\")[0]); // Include only the function name in the note
```
We'd suggest looking over this line - examples where there is a dot in the example, e.g. `document.cookie`, or `window.open` only show up in the notes as `document` or `window`, so it's hard to tell what the exact suspicious function call is.

BugBountyzip replied:
done

ps-porpoise commented on Proxy/HTTP/DetectSuspiciousJSFunctions.bambda:
```diff
@@ -0,0 +1,72 @@
+
+if (foundSuspiciousFunction) {
+    // Set the highlight color to RED and add notes
+    requestResponse.annotations().setHighlightColor(HighlightColor.RED);
```
We think highlighting should only be applied if the `manualColorHighlightEnabled` flag is set. Alternatively, you could remove this flag (see next review comment below).

ps-porpoise commented on Proxy/HTTP/DetectSuspiciousJSFunctions.bambda:
```diff
@@ -0,0 +1,72 @@
+    // Set the highlight color to RED and add notes
+    requestResponse.annotations().setHighlightColor(HighlightColor.RED);
+    if (manualColorHighlightEnabled && notesBuilder.length() > 0) {
+        requestResponse.annotations().setNotes("Suspicious JS functions detected: " + notesBuilder.toString());
```
You might want to rename the `manualColorHighlightEnabled` so it's clearer that it's also modifying notes.

ps-porpoise commented on Proxy/HTTP/DetectSuspiciousJSFunctions.bambda:
```diff
@@ -0,0 +1,72 @@
+ * Bambda Script to Detect and Highlight Suspicious JavaScript Functions
+ * It identifies a range of suspicious JavaScript functions often associated with unsafe practices or vulnerabilities, such as 'eval()', 'setTimeout()', and 'document.write()'.
+ * Upon detection, responses are highlighted in red, and notes are appended to indicate the specific functions found.
+ * Author: Tur24Tur
+ * GitHub: @BugBountyzip (https://github.com/BugBountyzip)
```
Please use `@author Tur24Tur / BugBountyzip (https://github.com/BugBountyzip)` instead here.

ps-porpoise commented on Proxy/HTTP/DetectSuspiciousJSFunctions.bambda:
```diff
@@ -0,0 +1,72 @@
+/**
+ * Bambda Script to Detect and Highlight Suspicious JavaScript Functions
+ * It identifies a range of suspicious JavaScript functions often associated with unsafe practices or vulnerabilities, such as 'eval()', 'setTimeout()', and 'document.write()'.
+ * Upon detection, responses are highlighted in red, and notes are appended to indicate the specific functions found.
```
Please move these 2 lines below the author tag.

BugBountyzip replied:
fixed

ps-porpoise reviewed (changes requested):
Thanks for your submission! We really like the concept, please find some feedback to improve it.

Let us know if you have any issues and we'll help you out!

BugBountyzip commented:
Thank you for your valuable feedback on the Bambda script. Your insights have been instrumental in enhancing the script's functionality and clarity. Please find below the updated code with the suggestions


```
/**
 * Bambda Script to Detect and Highlight Suspicious JavaScript Functions
  @author Tur24Tur / BugBountyzip (https://github.com/BugBountyzip)
  It identifies a range of suspicious JavaScript functions often associated with unsafe practices or vulnerabilities.
 * Upon detection, responses are highlighted in red and notes are appended, if enabled.
 **/

boolean enableManualAnnotations = true;

// Ensure there is a response
if (!requestResponse.hasResponse()) {
    return false;
}

// Check the Content-Type header for JavaScript
String contentType = requestResponse.response().headerValue("Content-Type");
if (contentType == null || !contentType.toLowerCase().contains("application/javascript")) {
    return false;
}

String responseBody = requestResponse.response().bodyToString();
boolean foundSuspiciousFunction = false;
StringBuilder notesBuilder = new StringBuilder();

// Expanded list of suspicious JavaScript functions
String[] suspiciousFunctions = {
    "eval\\(",                 // Executes a string as code
    "setTimeout\\(",           // Can execute strings as code if used improperly
    "setInterval\\(",          // Similar to setTimeout, can execute strings as code
    "document\\.write\\(",     // Can overwrite entire document
    "innerHTML",               // Can introduce XSS vulnerabilities if used with untrusted content
    "document\\.createElement\\(",  // Safe, but part of dynamic content generation which can be risky
    "document\\.execCommand\\(",   // Deprecated, was used to execute certain commands
    "document\\.domain",       // Altering the document.domain can be risky
    "window\\.location\\.href",    // Can be used for redirects which might be used in phishing
    "document\\.cookie",       // Accessing cookies can be sensitive
    "document\\.URL",          // Can be used to extract URL information
    "document\\.referrer",     // Can be used to check where the request came from
    "window\\.open\\(",        // Opening a new window or tab, potential for misuse
    "document\\.body\\.innerHTML", // Specific case of innerHTML, also risky
    "element\\.setAttribute\\(",   // If used improperly, can set risky attributes like 'onclick'
    "element\\.outerHTML",         // Similar risks to innerHTML
    "XMLHttpRequest\\(",           // Can be used for sending/receiving data, potential for misuse
    "fetch\\(",                    // Modern way to make network requests, potential for misuse
    "navigator\\.sendBeacon\\("    // Used to send analytics and tracking data
};

for (String function : suspiciousFunctions) {
    Pattern pattern = Pattern.compile(function);
    Matcher matcher = pattern.matcher(responseBody);
    if (matcher.find()) {
        foundSuspiciousFunction = true;
        if (enableManualAnnotations) {
            if (notesBuilder.length() > 0) {
                notesBuilder.append(", ");
            }
            notesBuilder.append(function); // Append the complete function signature
        }
    }
}

if (foundSuspiciousFunction && enableManualAnnotations) {
    requestResponse.annotations().setHighlightColor(HighlightColor.RED);
    if (notesBuilder.length() > 0) {
        requestResponse.annotations().setNotes("Suspicious JS functions detected: " + notesBuilder.toString());
    }
}

return foundSuspiciousFunction;


```

 1- Removed Redundant Null Check: The check for requestResponse.response() == null has been removed as hasResponse() already covers this scenario.

 2- Improved Function Name Extraction: Adjusted the logic to correctly extract and display the full function name in the notes, including cases like document.cookie and window.open.

 3- Conditional Highlighting Based on Flag: Highlighting now only applies if the enableManualAnnotations flag is set to true, ensuring that users have control over when highlights are applied.

4- Renamed Flag for Clarity: Renamed manualColorHighlightEnabled to enableManualAnnotations for better clarity, indicating that it controls both notes and highlighting features. and fixed the @autor name 


Thank you again

PortSwiggerWiener reviewed (approved):
Looks good 👍
